### PR TITLE
Dom builder ergonomics

### PR DESCRIPTION
**Pull request by simon-bourne**
This improve the DOM build ergonomics. See the new [`dom_builder` example](https://github.com/simon-bourne/moxie/blob/4152a99ebc6d1d4395c1c02691b3c993c80d0f5c/dom/examples/dom_builder/src/lib.rs) for details.

### CODE REVIEW

File: dom/examples/dom_builder/index.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<!doctype html>
+<html lang="en">
+
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=1" name="viewport" />
+    <title>moxie-dom • counter</title>
```
Comment (anp) [inline dom/examples/dom_builder/index.html:9]:
```suggestion
    <title>moxie-dom • builder</title>
```

File: mox/src/lib.rs [hunk 2]
```diff
@@ -12,11 +12,11 @@ use syn::{
 };
 use syn_rsx::{NodeName, NodeType};
 
-/// Accepts an XML-like expression and expands it to builder method calls.
```
Comment (anp) [inline mox/src/lib.rs:15]: I think this is still accurate, maybe it should say "builder-like method calls"?

File: mox/src/lib.rs [hunk 3]
```diff
@@ -347,8 +342,9 @@ impl ToTokens for MoxAttr {
     }
 }
 
-struct MoxExpr {
-    expr: syn::Expr,
+enum MoxExpr {
```
Comment (anp) [inline mox/src/lib.rs:345]: I'm not sure I follow what this is for -- is it only to add `#[allow(unused_braces)]` on some blocks?

File: index.html [hunk 4]
```diff
@@ -180,6 +180,10 @@ <h4>(toy) examples</h4>
                 <a href="dom/examples/counter_struct/index.html">Counter (struct w/ traits)</a>
                 (<a href="https://github.com/anp/moxie/tree/main/dom/examples/counter_struct">source</a>)
             </li>
+            <li>
+                <a href="dom/examples/dom_builder/index.html">Counter (bare function) using DOM builder API</a>
```
Comment (anp) [inline index.html:184]:
```suggestion
                <a href="dom/examples/dom_builder/index.html">Counter (no-XML DOM builder API)</a>
```

Comment (anp) [reply index.html:184]: How does that sound to you?

File: dom/src/interfaces/node.rs [hunk 5]
```diff
@@ -44,26 +44,24 @@ pub trait NodeWrapper: sealed::Memoized + Sized {
...
-        self
-    }
 }
 
-/// Allows values which `impl Display` to be used directly as `mox!` children,
-/// converting them into text nodes. Can be implemented for your type.
-pub trait TextChild: Sized {
-    /// Wrap self into the `text` node
-    fn into_child(self) -> Text;
+/// A builder for DOM nodes
```
Comment (anp) [inline dom/src/interfaces/node.rs:49]:
```suggestion
/// A type which can be built into a DOM node.
```

File: dom/src/interfaces/node.rs [hunk 6]
```diff
@@ -44,26 +44,24 @@ pub trait NodeWrapper: sealed::Memoized + Sized {
...
+pub trait NodeBuilder {
+    /// The type of the DOM node
+    type Target;
```
Comment (anp) [inline dom/src/interfaces/node.rs:52]:
```suggestion
    type Output;
```

?

**Review by anp — COMMENTED** [review]
Looks solid! A few nits and questions.

**Comment by anp** [comment]
If you rebase onto main again you should be able to get meaningful results from CI.

File: mox/src/lib.rs (hunk 3 above)
Comment (simon-bourne) [reply mox/src/lib.rs:345]: Exactly. Let me know if you think it warrants a comment or you'd rather just blanket `#[allow(unused_braces)]`.

Comment (anp) [reply mox/src/lib.rs:345]: Yeah, let's just slap the unused braces lint on both options if we need to.

File: mox/src/lib.rs (hunk 2 above)
Comment (simon-bourne) [reply mox/src/lib.rs:15]: Fixed

File: mox/src/lib.rs (hunk 3 above)
Comment (simon-bourne) [reply mox/src/lib.rs]: Fixed

File: dom/src/interfaces/node.rs (hunk 6 above)
Comment (simon-bourne) [reply dom/src/interfaces/node.rs]: Fixed

**Comment by simon-bourne** [comment]
I applied all your suggestions in the comments, and all non-codecov checks are passing. The only actual reduction in code coverage I can see, is not including the "identity" `impl NodeBuilder for Text`. I felt that would be adding a test for the sake of coverage, but happy to add one if you think otherwise.

If you're happy with the coverage, and there's no other fixes, I think we're good to go.

Here's the code not covered by tests:

```rust
impl NodeBuilder for Text {
    type Target = Self;

    fn build(self) -> Self::Target {
        self
    }
}
```

**Comment by anp** [comment]
Awesome, thanks!

**Pull request merged**